select repository and sync with url

tatchi opened the pull request:
This PR allows selecting a repository (via a `<select />`) and sync the URL based on the selected repo. I guess the next step would be to filter data based on the selected date ranges and sync that with the URL too. 

https://user-images.githubusercontent.com/5595092/107287923-f41fdc00-6a62-11eb-8360-2a71b777ff06.mov

We currently fetch all the data from the database and filter them out on the frontend. I'm planning to explore if we couldn't make that better by letting some components fetching their own data.

tatchi commented on frontend/src/Litepicker.res:
```diff
@@ -47,7 +46,7 @@ let make = (~sx as uSx=[], ~startDate=?, ~endDate=?, ~onSelect=?) => {
     let _self = init(options)
 
     None
-  }, [dates])
```
Looks like this `dates` value was not used

rizo replied:
You're right. I honestly don't remember what I was trying to do here.

rizo commented on frontend/src/App.res:
```diff
@@ -18,19 +18,44 @@ query ($startDate: timestamp!, $endDate: timestamp!) {
-  -compare(pn1, pn2)
+module PullCompare = Belt.Id.MakeComparable({
+  type t = (int, option<string>)
+  let cmp = compare
```
Note that this seems to drop the `-` which was there to unsure that the PRs are listed from the most recent to the oldest.

tatchi replied:
Oops, that was not intentional. Good catch 👍

rizo commented on frontend/src/Sidebar.res:
```diff
@@ -11,30 +11,51 @@ let pullToString = ((pull_number, branch)) =>
-    ->Rx.array}
-  </Column>
+module PullsMenu = {
+  @react.component
```
I'm never sure when something needs to be an actual React component or just a function producing an element.

My convention so far has been to create a new component if:
1. The view needs to manage its own state.
2. The view is expensive to render.

But I have no idea if this has any meaningful implications. Do you have any thoughts on this?

tatchi replied:
I think that makes sense. As long as you don't use `state`, `useEffect`, `memo` et al, both should be fine (I'm not sure it does matter much in this case)

rizo commented on frontend/src/Sidebar.res:
```diff
@@ -11,30 +11,51 @@ let pullToString = ((pull_number, branch)) =>
-      </Row>
-    </a>
-    {pullsMenu(~pulls)}
+    <select
```
Can we use the `Components.Select` here? I'm not very happy with all the components, but have plans to improve them at some point. For now would be good to keep the UI somewhat consistent.

tatchi replied:
I should have checked if something was already existing for that. Thanks!

rizo commented on frontend/src/Sidebar.res:
```diff
@@ -11,30 +11,51 @@ let pullToString = ((pull_number, branch)) =>
+let make = (
+  ~pulls,
+  ~selectedRepo,
+  ~repos,
```
The backend and the db use `repo_id` to represent `$org_name/$repo_name`. Can we keep this convention here and call this prop `repo_ids` instead of `repos`?

rizo commented on frontend/src/BenchmarkTest.res:
```diff
@@ -7,10 +7,10 @@ type testMetrics = {
 
-let commitUrl = commit => `https://github.com/mirage/index/commit/${commit}`
-let goToCommitLink = commit => {
+let commitUrl = (repo, commit) => `https://github.com/${repo}/commit/${commit}`
```
Similar to my comment for the sidebar, can we use `repo_id` instead of `repo`. My suggestion is to always use `repo_id` to represent `repo_org/repo_name`.

tatchi replied:
Makes sense. On a side note, we're not consistent with our case style. Some variables are `snake_case` (mainly fields from GraphqQL) and the others `camelCase`

rizo replied:
Yes, I agree those should be consistent. It's all my fault, I've been switching a lot between OCaml and ReScript projects and have a tendency to use underscores without realising it. Please feel free to point this out in reviews.

rizo commented on frontend/src/App.res:
```diff
@@ -18,19 +18,44 @@ query ($startDate: timestamp!, $endDate: timestamp!) {
-let collectPulls = (data: array<GetBenchmarks.t_benchmarks>): array<(int, option<string>)> => {
-  let data = Belt.List.fromArray(data)
-  let data = Belt.List.keepMap(data, (item: GetBenchmarks.t_benchmarks) =>
+let collectPullsForRepo = (~repo, benchmarks: array<GetBenchmarks.t_benchmarks>): array<(
```
I think this is ok for now while we have two repos. Once we have more we should pass the `repo_id` to the GraphQL query.

Happy to merge this as is, just something to keep in mind for the follow up work.

rizo commented on frontend/src/App.res:
```diff
@@ -18,19 +18,44 @@ query ($startDate: timestamp!, $endDate: timestamp!) {
+  ->Belt.Set.toArray
+}
+
+let collectRepos = (benchmarks: array<GetBenchmarks.t_benchmarks>): array<string> => {
```
On a second thought, why not use GraphQL now? It seems to complicate quite a bit data filtering. Would it be possible to make another query to obtain the distinct list of repositories?

tatchi replied:
I agree it would be better to query for specific data at different places instead of querying everything and filtering on the frontend side. I would prefer to address that in another PR though if that's ok for you?

rizo replied:
That's fine, of course.

rizo commented on frontend/src/App.res:
```diff
@@ -131,35 +200,75 @@ let make = () => {
+                <div> {"This pull request does not exist!"->Rx.string} </div>
+              }
+            }
+          | _ =>
```
This shouldn't be `_` as that will capture anything except `/pull/xxx`. I think what we want is `list{}`?

tatchi replied:
Indeed, `_` is less restrictive since anything after `org/repo` other than `/pull/xxx` (e.g `org/repo/whatever/else`) would still display data for the repo (instead of showing an "unknown route error").

rizo reviewed (commented):
Left some suggestions on data fetching/processing. Everything else looks pretty good! 👍

tatchi commented:
@rizo thanks for your review, I should have addressed all your comments except for the GraphQL stuff. I would prefer to handle it in another PR if that's fine for you.

rizo commented:
Thanks, @tatchi!
